lib/vrl/stdlib/src/parse_query_string.rs

```diff
@@ -0,0 +1,81 @@
+use std::collections::BTreeMap;
+use url::form_urlencoded;
+use vrl::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct ParseQueryString;
+
+impl Function for ParseQueryString {
+    fn identifier(&self) -> &'static str {
+        "parse_query_string"
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[Example {
+            title: "parse query string",
+            source: r#"parse_query_string!("foo=1&bar=2")"#,
+            result: Ok(r#"
+                {
+                    "foo": "1",
+                    "bar": "2"
+                }
+            "#),
+        }]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Compiled {
+        let value = arguments.required("value");
+        Ok(Box::new(ParseQueryStringFn { value }))
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "value",
+            kind: kind::ANY,
+            required: true,
+        }]
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ParseQueryStringFn {
+    value: Box<dyn Expression>,
+}
+
+impl Expression for ParseQueryStringFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let query_string = self.value.resolve(ctx)?.try_bytes()?;
+        let result = form_urlencoded::parse(query_string.as_ref())
+            .map(|(k, v)| (k.to_string(), v.into()))
+            .collect();
```

Comment:
Our other parse functions will return an error if it can't actually parse any data. I think we should do the same here.

What does `form_urlencoded::parse` return if we pass a string that isn't `key=value`, eg `form_urlencoded::parse("thing")`? Does it ignore that data? Does it become the key with an empty value?

If it does ignore that data, then I think we should raise an error here, so if `result.len() == 0` for any parameter other than an empty string, we should error.

Reply:
Good point! It returns a key with an empty value, so I added more tests and made this function `infallible`. Also updated the doc.

Reply:
Just circling back to this. To @FungusHumungus's point, we've typically had the `parse_` functions (like `parse_grok`) fail if the provided input doesn't match the pattern. I could see doing the same here and returning an error if given an input that is not a valid query string.

Reply:
@jszwedko I added more tests(and updated the doc) to show that we support empty keys and values(I believe empty values must be allowed and empty keys can be filtered by users with `compact`). Given that, I'm struggling to think of any invalid input, therefore I made this function `infallible`. Do you see any example of invalid input, which would require error handling?

Reply:
👍 I didn't realize empty values were allowed. That makes sense.